src/wasm/nxt_wasm.c

```diff
@@ -0,0 +1,262 @@
+        read_bytes = NXT_WASM_MEM_SIZE - offset;
+    }
+
+    bytes_read = nxt_unit_request_read(req, (uint8_t *)wr + offset, read_bytes);
+    wr->content_sent = wr->total_content_sent = content_sent = bytes_read;
+
+    wr->request_size = offset + bytes_read;
+
+    nxt_wasm_ctx.req = req;
+    nxt_wops->exec_request(&nxt_wasm_ctx);
+
+    if (content_len == content_sent) {
+        goto request_done;
+    }
+
+    wr->nr_fields = 0;
+    wr->content_offs = offset = sizeof(nxt_wasm_request_t);
+    for ( ; ; ) {
+        read_bytes = nxt_min(content_len - content_sent,
+                             NXT_WASM_MEM_SIZE - offset);
+        bytes_read = nxt_unit_request_read(req, (uint8_t *)wr + offset,
+                                           read_bytes);
+
+        content_sent += bytes_read;
+        wr->request_size = wr->content_sent = bytes_read;
+        wr->total_content_sent = content_sent;
+
+        nxt_wops->exec_request(&nxt_wasm_ctx);
+
+        if (content_len == content_sent) {
+            break;
+        }
+    }
+
+request_done:
+    NXT_WASM_DO_HOOK(NXT_WASM_FH_REQUEST_END);
+}
+
+
+static nxt_int_t
+nxt_wasm_start(nxt_task_t *task, nxt_process_data_t *data)
+{
+    nxt_int_t              ret;
+    nxt_unit_ctx_t         *unit_ctx;
+    nxt_unit_init_t        wasm_init;
+    nxt_common_app_conf_t  *conf;
+
+    conf = data->app;
+
+    ret = nxt_unit_default_init(task, &wasm_init, conf);
+    if (nxt_slow_path(ret != NXT_OK)) {
+        nxt_alert(task, "nxt_unit_default_init() failed");
+        return ret;
+    }
+
+    wasm_init.callbacks.request_handler = nxt_wasm_request_handler;
+
+    unit_ctx = nxt_unit_init(&wasm_init);
+    if (nxt_slow_path(unit_ctx == NULL)) {
+        return NXT_ERROR;
+    }
+
+    NXT_WASM_DO_HOOK(NXT_WASM_FH_MODULE_INIT);
+    nxt_unit_run(unit_ctx);
+    nxt_unit_done(unit_ctx);
+    NXT_WASM_DO_HOOK(NXT_WASM_FH_MODULE_END);
+
+    nxt_wops->destroy(&nxt_wasm_ctx);
+
+    exit(EXIT_SUCCESS);
+}
+
+
+static nxt_int_t
+nxt_wasm_setup(nxt_task_t *task, nxt_process_t *process,
+               nxt_common_app_conf_t *conf)
+{
+    int                      err;
+    nxt_wasm_app_conf_t      *c;
+    nxt_wasm_func_handler_t  *fh;
+
+    c = &conf->u.wasm;
+
+    nxt_wops = &wasm_ops;
+
+    nxt_wasm_ctx.module_path = c->module;
+
+    fh = nxt_wasm_ctx.fh;
+
+    fh[NXT_WASM_FH_REQUEST].func_name = c->request_handler;
+    fh[NXT_WASM_FH_MALLOC].func_name = c->malloc_handler;
+    fh[NXT_WASM_FH_FREE].func_name = c->free_handler;
+
+    /* Optional function handlers (hooks) */
+    fh[NXT_WASM_FH_MODULE_INIT].func_name = c->module_init_handler;
+    fh[NXT_WASM_FH_MODULE_END].func_name = c->module_end_handler;
+    fh[NXT_WASM_FH_REQUEST_INIT].func_name = c->request_init_handler;
+    fh[NXT_WASM_FH_REQUEST_END].func_name = c->request_end_handler;
+    fh[NXT_WASM_FH_RESPONSE_END].func_name = c->response_end_handler;
+
+    err = nxt_wops->init(&nxt_wasm_ctx);
+    if (err) {
```

Comment:
You'll probably need to compare `== NXT_ERROR` (or `!= NXT_OK` if other values are possible), right?